[docs] Simplify loading guide

stevhliu opened the pull request:
This PR moves the high-level explanations in the loading guide into a separate Conceptual Guide so users can focus on executing the code more quickly. The explanatory guide is linked for users who are interested in learning more about what's happening.

stevhliu commented on docs/source/en/using-diffusers/loading.mdx:
````diff
@@ -191,349 +122,50 @@ model_id = "runwayml/stable-diffusion-v1-5"
-pipe.save_pretrained("./stable-diffusion-v1-5", variant="fp16")
-```
-
-If we don't save into an existing `stable-diffusion-v1-5` folder the new folder would look as follows:
````
Does this mean you should save it into an existing folder or repository like `runway/stable-diffusion-v1-5` for example?

sayakpaul replied:
If we are saving it locally (which is what `save_pretrained()` is doing), then how would it matter?

yiyixuxu replied:
I was confused when I was reading this part too!

I think what he meant might be that you should save it into a new folder in order to follow this particular example (not because it is better to save it into a new folder)

-i.e. if you save it into an existing folder instead, the folder structure will look different than what's show below (it will contain both  .fp16.bin extension and .bin., not just fp16.bin) and you will NOT get an Exception error if you load without `variant` argument

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_2694). All of your documentation changes will be reflected on that endpoint.

stevhliu commented on docs/source/en/using-diffusers/loading.mdx:
````diff
@@ -191,349 +122,50 @@ model_id = "runwayml/stable-diffusion-v1-5"
-
-
-```py
-DiffusionPipeline.from_pretrained("./stable-diffusion-v1-5", variant="fp16", torch_dtype=torch.float16)
````
Do you need to specify `torch_dtype=torch.float16` if you're already specifying the `variant=fp16` or is this just the case because you don't have the `main` checkpoints saved here?

patrickvonplaten replied:
You need to specify it - these are two different things

stevhliu commented on docs/source/en/using-diffusers/loading.mdx:
```diff
@@ -191,349 +122,50 @@ model_id = "runwayml/stable-diffusion-v1-5"
-#### How to save and load variants
-
-Saving a diffusion pipeline as a variant can be done by providing [`DiffusionPipeline.save_pretrained`] with the `variant` argument. 
-The `variant` extends the weight name by the provided variation, by changing the default weight name from `diffusion_pytorch_model.bin` to `diffusion_pytorch_model.{variant}.bin` or from `diffusion_pytorch_model.safetensors` to `diffusion_pytorch_model.{variant}.safetensors`. By doing so, one creates a variant of the pipeline checkpoint that can be loaded **instead** of the "main" pipeline checkp ...
```
Why doesn't the `runwayml/stable-diffusion-v1-5` fp16 unet [checkpoint](https://huggingface.co/runwayml/stable-diffusion-v1-5/tree/fp16/unet) have the `variant` in the weight name?

patrickvonplaten replied:
I think we're at a stage where we can/should add it. Will upload all the variants this week

stevhliu reviewed (commented):
This is just a draft because I have some questions about `variants` :)

sayakpaul commented on docs/source/en/conceptual/pipeline_explained.mdx:
```diff
@@ -0,0 +1,197 @@
+
+In addition to the original pipeline checkpoints stored in a repository, there may also be *checkpoint variants*. A variant is typically checkpoint weights stored in a lower precision and lower storage data type like `fp16` or they may be non-exponential mean averaged (EMA) weights so you can resume finetuning from a checkpoint. Variants are advantageous in specific scenarios - half-precision che ...
+
+This means other serialization formats, such as [Safetensors](./using-diffusers/using_safetensors), are not considered checkpoint variants because their weights are identical to the original checkpoint. It may also be tempting to consider different model structures as variants, such as [`stable-diffusion-v1-5`](https://huggingface.co/runwayml/stable-diffusion-v1-5) and [`stable-diffusion-2-0`](ht ...
```
```suggestion
This means other serialization formats, such as [Safetensors](./using-diffusers/using_safetensors), are not considered checkpoint variants because their weights are identical to the original checkpoint. It may also be tempting to consider different model structures as variants, such as [`stable-diffusion-v1-5`](https://huggingface.co/runwayml/stable-diffusion-v1-5) and [`stable-diffusion-2`](https://huggingface.co/stabilityai/stable-diffusion-2). However, these checkpoints aren't considered variants because `stable-diffusion-v1-5` uses a different `CLIPTextModel` than `stable-diffusion-2`.
```

sayakpaul commented on docs/source/en/using-diffusers/loading.mdx:
```diff
@@ -10,20 +10,26 @@ an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express o
-- *Complete Diffusion Pipelines* via the [`DiffusionPipeline.from_pretrained`]
-- *Diffusion Models* via [`ModelMixin.from_pretrained`]
-- *Schedulers* via [`SchedulerMixin.from_pretrained`]
+- pipelines from the Hub and locally
```
```suggestion
- Pipelines from the Hub and locally
```

sayakpaul reviewed (approved):
Nice cleanup!

I really like the new "explained" doc. But I would request @patrickvonplaten to take a closer look in case I missed anything.

yiyixuxu commented on docs/source/en/conceptual/pipeline_explained.mdx:
```diff
@@ -0,0 +1,197 @@
+specific language governing permissions and limitations under the License.
+-->
+
+# Pipelines explained
```
A different title? This guide explains one aspect of the pipeline regards to loading. Maybe "`DiffusionPipeline` explained"? ( `DiffusionPipeline` is the base model that's responsible for these tasks that you are talking about here)

patrickvonplaten commented on docs/source/en/conceptual/pipeline_explained.mdx:
```diff
@@ -0,0 +1,197 @@
+
+### Checkpoint variants
+
+In addition to the original pipeline checkpoints stored in a repository, there may also be *checkpoint variants*. A variant is typically checkpoint weights stored in a lower precision and lower storage data type like `fp16` or they may be non-exponential mean averaged (EMA) weights so you can resume finetuning from a checkpoint. Variants are advantageous in specific scenarios - half-precision che ...
```
Let's also link to: https://huggingface.co/docs/diffusers/using-diffusers/loading#loading-variants

patrickvonplaten commented on docs/source/en/using-diffusers/loading.mdx:
````diff
@@ -191,349 +122,50 @@ model_id = "runwayml/stable-diffusion-v1-5"
-stable_diffusion_img2img = StableDiffusionImg2ImgPipeline(**components)
 ```
 
-Note how the above code snippet makes use of [`DiffusionPipeline.components`].
````
Wait why do you want to remove this here?

patrickvonplaten commented:
Hmm not too happy about the changes here. I see the benefit of simplifying the loading doc - however all the info that is deleted is very important for users. I don't think we should do a split (loading / conceptual guide here). I think we should leave everything on the loading page (which is visited a lot) and just split everything into "basic" and "advanced" use cases. 

I don't think it makes sense to add a now "explanatory guide" here. IMO many users just want a quick look-up they can visit quickly to see how variants work instead of having to read through a guide/tutorial. 

I see `loading` much closer to API doc string that is looked up a lot instead of something like a tutorial.

stevhliu commented:
Thanks for the feedback!

I've refactored to keep everything on one page (I'll delete the "explained" page once we're happy with this), and split it into practical stuff at the top and explanatory things at the bottom. I think the practical steps are important for the quick look-up without necessarily needing to know how the folder is structured and the details of the `model_index.json` file, for example. 

But if they are interested, I added a `<Tip>` where they can skip to the end and get more context about how it works.

I think all the info should still be on the page, but feel free to let me know if I'm missing anything!

patrickvonplaten commented on docs/source/en/using-diffusers/loading.mdx:
```diff
@@ -191,227 +132,189 @@ model_id = "runwayml/stable-diffusion-v1-5"
-A checkpoint stored in [torch's half-precision / float16 format](https://pytorch.org/blog/accelerating-training-on-nvidia-gpus-with-pytorch-automatic-mixed-precision/) requires only half the bandwith and storage when downloading the checkpoint,
-**but** cannot be used when continuing training or when running the checkpoint on CPU.
-Similarly the *non-exponential-averaged* (or non-EMA) version of the checkpoint should be used when continuing fine-tuning of the model checkpoint, **but** should not be used when using the checkpoint for inference. 
+Load a variant by specifying the `variant` argument in [`DiffusionPipeline.from_pretrained`], and the `torch_dtype` if the variant is a different floating point type. 🧨 Diffusers won't download a variant unless it is explicitly specified, so you don't have to worry about downloading and caching more checkpoints than you need.
```
Here we need to be more precise I think.

`variant` -> this just defines what files should be loaded
`torch_dtype` -> this defines into what `torch_dtype` the weights should be converted to. 

Can we maybe try to say something along the following:
```suggestion
Load a variant by specifying the `variant` argument in [`DiffusionPipeline.from_pretrained`]. 🧨 Diffusers won't download a variant unless it is explicitly specified, so you don't have to worry about unnecessarily downloading and caching more checkpoints than you need.

 The `torch_dtype` argument is unrelated to the `variant` argument and decides which floating point precision the loaded checkpoints should have. If you want to save bandwidth by loading a `"fp16"` variant, you should also define `torch_dtype=torch.float16` as otherwise the fp16 weights will be converted to the default fp32 precision. Note that you can also load the original checkpoint without setting a variant and then converting it to float16 precision by passing `torch_dtype=torch.float16`. In this case you downloaded float32 weights and converted them to float16 after loading.
```

patrickvonplaten replied:
It would be great if the reader could understand the difference between `torch_dtype` and `variant` here (they are completely orthogonal.

stevhliu replied:
Thanks for the explanation, this is super helpful!

patrickvonplaten reviewed (commented):
Looks nice! Could we just add some more specification about the difference between variant and torch_dtype?

patrickvonplaten reviewed (approved):
Looks good to me!
